django/db/models/sql/compiler.py

```diff
@@ -1524,6 +1545,20 @@ def apply_converters(self, rows, converters):
                 row[pos] = value
             yield row
 
+    def has_composite_fields(self, expressions):
+        # Check for composite fields before calling the relatively costly
+        # composite_fields_to_tuples.
+        return any(isinstance(expression, ColPairs) for expression in expressions)
+
+    def composite_fields_to_tuples(self, rows, expressions):
+        for row in map(list, rows):
+            for i, expression in enumerate(expressions):
+                if isinstance(expression, ColPairs):
+                    pos = slice(i, i + len(expression))
+                    row[pos] = (tuple(row[pos]),)
+
+            yield row
```

Comment:
I think we might be able to optimise this a bit:

```suggestion
        col_pair_slices = [
            slice(i, i + len(expression))
            for i, expression in enumerate(expressions)
            if isinstance(expression, ColPairs)
        ]
        for row in map(list, rows):
            for pos in col_pair_slices:
                row[pos] = (tuple(row[pos]),)

            yield row
```

This avoids iterating over all `expressions` for every row. Instead we can calculate the relevant slices to update upfront and then apply them to each row.

Reply:
Nice! Thanks!